### PR TITLE
Restructured BotKillerAgent.cs

### PR DESCRIPTION
**Pull request by SiennaRose**
Broke CollectObservations() into smaller methods to make the different concepts more understandable: 
- ObserveHealth()
   -->UpdateTrackedAgentHealthAndRewardAgent()
- ObserveAmmo()
- ObserveEnemies()
   -->GetEnemyObservations()
   -->ObserveEnemyHealth()
   -->UpdateTrackedEnemyHealthAndRewardAgent()
- ObservePickups()
   -->GetPickupLocation()

Made distance, bearing, heading, and line-of-sight variables into their own methods b/c some of them were being reused to get the distance and bearing of both pickups and enemies:
- DoesAgentHaveLineOfSightOf()
    --> AreObstaclesBetweenAgentAnd()
- DetermineDistanceFrom()
- DetermineBearingTo()
- DetermineHeadingOf()

Removed Commented Out (Dead) Code

### CODE REVIEW

File: Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs
Comment (zako42) [inline Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs:88]: This is a good refactor into methods

File: Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs
Comment (zako42) [inline Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs:97]: since we have the `using Unity.FPS.Game` we can should be able to call `GetComponent<Health>()` instead of the fully qualified name.  Doesn't hurt to be specific, but you can also save some extra typing.  Can go either way

File: Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs
Comment (zako42) [inline Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs:103]: normally we'd like to have a method do 1 task, this is doing 2 (update and reward)
but at the same time, the way we handle observations and rewards, it's best to keep them together so it makes sense

File: Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs
Comment (SiennaRose) [reply Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs:88]: Thanks 😁

File: Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs
Comment (zako42) [inline Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs:105]: one way you can simplify this is to use a "guard clause"
since all the code is nested inside an if statement, you can use the opposite condition to exit out early.

so `if (agentHealth == trackedHealth) return`
then the rest of the code doesn't need to be nested, and there's less cognitive load
(cognitive load is a real thing -- when dealing with production projects it becomes more important)

File: Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs
Comment (zako42) [inline Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs:174]: we should put in a conditional to make sure we dont have a null Health component.  if null, then it can skip and use 0

File: Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs
Comment (SiennaRose) [reply Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs]: Replaced 😋

File: Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs
Comment (SiennaRose) [reply Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs]: split into different methods UpdateAgentTrackedHealth() and DisplayHealthReward() 🤠

File: Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs
Comment (SiennaRose) [reply Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs]: On Guard! 🤺

File: Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs
Comment (SiennaRose) [reply Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs:177]: Fixed! you null it 🤩

File: Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs
Comment (zako42) [inline Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs:112]: Since this method also does the actual rewarding using `AddReward()` we should rename this method to something like "RewardAgent", just since the name might be confusing (people may think it's just doing the logging, not the actual rewards).  Since it calls the `DebugLogReward()` method, maybe something we could do in the BotKiller class, we can have a public bool for "debugOutput" (default to false) and have something in the method where it checks the debugOuput.  If true, then do the `DebugLogReward()` but if false, then skip it.

One thing you could also try would be to _override_ the actual `AddReward()` method (part of unity's agent class).  In the overrided AddReward in BotKiller, you would call `base.AddReward(value)` and also add in the `DebugLogReward()`.  That could be clean.  Would need to replace the calls to this DisplayReward method and use the overriden AddReward instead.

I'll leave it up to you to decide what you think is best!  You can test it out and udpate however you'd like to proceed

Comment (SiennaRose) [reply Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs]: went w/ numero uno! 😉👍

**Review by zako42 — APPROVED** [review]
looks good!  i can merge right after this

**Comment by zako42** [comment]
ok merged!  this cleans up a lot of the bot killer script and should be much easier for people to look at and figure out whats going on.

**Pull request merged**
